### PR TITLE
Graph refactor

### PR DESCRIPTION
**Pull request by karliss**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

This continuation of what was started #17362 in.

* Support all graph output formats for `acg`
* Convert `agx` and `agi` to use the same mechanism

There are a bunch more more or less manual graph printing that could be refactored this way, but I wanted to to keep first PR smaller and discuss the concept of changes before refactoring everything.

Goal of changes (within this PR and follow up PR) are:
* reduce usage of global state for implementing commands (core->graph)
* have a C api for obtaining drawable graph structure that is somewhere between exact drawing and Ranal structures.
* Include information object's addres when it makes sense so that it can be used for interactive functionality. In many of the r2 graphs each node represents something that have address.
* Use agraph only when drawing a graph in terminal.

@ret2libc Had a strong opinion related to where exactly this code should be located. This might be a good chance to discuss that further.

**Test plan**

* r2r tests work
* Interactive graph modes agiv, agxv, acgv work

**Closing issues**

...

### CODE REVIEW

File: libr/core/cmd_anal.c [hunk 1]
```diff
@@ -8058,35 +8058,71 @@ static char *getViewerPath(void) {
 	return NULL;
 }
 
-static char* graph_cmd(RCore *core, char *r2_cmd, const char *save_path) {
+static char *dot_executable_path() {
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8061]: tests say this should be `...(void)`

File: libr/util/graph.c [hunk 2]
```diff
@@ -162,6 +162,16 @@ R_API RGraphNode *r_graph_add_node(RGraph *t, void *data) {
 	return n;
 }
 
+R_API RGraphNode *r_graph_add_nodef(RGraph *graph, void *data, RListFree user_free) {
+	RGraphNode *node = r_graph_add_node (graph, data);
+	if (node) {
+		node->free = user_free;
+	} else if (user_free) {
+		user_free (data);
```
Comment (ret2libc) [inline libr/util/graph.c:170]: I'm not sure about this. If the allocation goes wrong, `data` should still be owned by the caller, no?
The caller would do something like:
```C
n = r_graph_add_nodef (g, data, ufree);
if (!n) {
   ufree (data);
   // free other stuff defined in this function
}
```

File: libr/util/graph.c [hunk 3]
```diff
@@ -316,6 +328,23 @@ R_API RGraphNodeInfo *r_graph_create_node_info(char *title, char *body, ut64 off
 		data->title = title;
 		data->body = body;
 		data->offset = offset;
+	} else {
+		free (title);
```
Comment (ret2libc) [inline libr/util/graph.c:332]: same here. Do we really want these free here? Maybe I'm wrong, but I think the "ownership" should be transferred only when everything goes right. If something goes wrong the objects still belong to the user, who can still do whatever it wants with those data.

File: libr/util/graph.c [hunk 4]
```diff
@@ -316,6 +328,23 @@ R_API RGraphNodeInfo *r_graph_create_node_info(char *title, char *body, ut64 off
...
 	}
 	return data;
-}
\ No newline at end of file
+}
+
+R_API RGraphNode *r_graph_add_node_info(RGraph *graph, char *title, char *body, ut64 offset) {
+	RGraphNodeInfo *data;
+	if (!graph) {
+		free (title);
```
Comment (ret2libc) [inline libr/util/graph.c:341]: same ownership thing.

File: libr/core/agraph.c [hunk 5]
```diff
@@ -4939,6 +4939,7 @@ R_API RAGraph *create_agraph_from_graph(const RGraph/*<RGraphNodeInfo>*/ *graph)
 	if (!result_agraph) {
 		return NULL;
 	}
+	result_agraph->need_reload_nodes = false;
```
Comment (ret2libc) [inline libr/core/agraph.c:4942]: Why was this required?

File: libr/include/r_core.h [hunk 6]
```diff
@@ -578,16 +578,16 @@ R_API void r_core_anal_esil(RCore *core, const char *str, const char *addr);
...
 R_API int r_core_anal_search(RCore *core, ut64 from, ut64 to, ut64 ref, int mode);
 R_API int r_core_anal_search_xrefs(RCore *core, ut64 from, ut64 to, int rad);
-R_API int r_core_anal_data (RCore *core, ut64 addr, int count, int depth, int wordsize);
+R_API int r_core_anal_data(RCore *core, ut64 addr, int count, int depth, int wordsize);
 R_API void r_core_anal_datarefs(RCore *core, ut64 addr);
 R_API void r_core_anal_coderefs(RCore *core, ut64 addr);
-R_API void r_core_anal_codexrefs(RCore *core, ut64 addr);
-R_API void r_core_anal_importxrefs(RCore *core);
+R_API RGraph/*<RGraphNodeInfo>*/ *r_core_anal_codexrefs(RCore *core, ut64 addr);
+R_API RGraph /*<RGraphNodeInfo>*/ *r_core_anal_importxrefs(RCore *core);
```
Comment (ret2libc) [inline libr/include/r_core.h:590]:
```suggestion
R_API RGraph/*<RGraphNodeInfo>*/ *r_core_anal_importxrefs(RCore *core);
```

**Review by ret2libc — COMMENTED** [review]
Some initial comments, I still need to go through the whole thing.

File: libr/core/cmd_anal.c [hunk 7]
```diff
@@ -8286,25 +8323,248 @@ R_API void r_core_agraph_print (RCore *core, int use_utf, const char *input) {
...
+		case 'k': // "ag_k"
+		{
+			Sdb *db = r_agraph_get_sdb (agraph);
+			char *o = sdb_querys (db, "null", 0, "*");
+			r_cons_print (o);
+			free (o);
+			break;
+		}
+		case 'v': // "ag_v"
+		case 'i': // "ag_i" - open current core->graph in interactive mode
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8489]: Is this missing from the help? I see `v Interactive ascii art` but I can't find `i`.

File: libr/core/cmd_anal.c [hunk 8]
```diff
@@ -8286,25 +8323,248 @@ R_API void r_core_agraph_print (RCore *core, int use_utf, const char *input) {
...
+		{
+			RANode *ran = r_agraph_get_first_node (agraph);
+			if (ran) {
+				ut64 oseek = core->offset;
+				r_agraph_set_title (agraph, r_config_get (core->config, "graph.title"));
+				r_agraph_set_curnode (agraph, ran);
+				agraph->force_update_seek = true;
+				agraph->need_set_layout = true;
+				agraph->layout = r_config_get_i (core->config, "graph.layout");
+				int ov = r_cons_is_interactive ();
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8499]:
```suggestion
				bool ov = r_cons_is_interactive ();
```

File: libr/core/cmd_anal.c [hunk 9]
```diff
@@ -8623,25 +8871,21 @@ static void cmd_anal_graph(RCore *core, const char *input) {
...
                                 char *cmdargs = r_str_newf ("agdd 0x%"PFMT64x, core->offset);
-                                char *cmd = graph_cmd (core, cmdargs, input + 2);
-                                if (cmd && *cmd) {
-                                        r_core_cmd0(core, cmd);
-                                }
-                                free(cmd);
-                                free(cmdargs);
-                                break;
+                                convert_dotcmd_to_image (core, cmdargs, input + 2);
+				free (cmdargs);
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8887]: tabs

File: libr/core/cmd_anal.c [hunk 10]
```diff
@@ -8058,35 +8058,71 @@ static char *getViewerPath(void) {
...
-	return cmd;
+	r_core_cmd0 (core, cmd);
+	result = true;
+end:
+	free (cmd);
+	free (dot);
+	return result;
+}
+
+static bool convert_dotcmd_to_image(RCore *core, char *r2_cmd, const char *save_path) {
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8107]: r2_cmd is never used

File: test/db/cmd/cmd_ag [hunk 11]
```diff
@@ -0,0 +1,337 @@
+NAME=agi
+FILE=bins/elf/hello_world
+CMDS=<<EOF
+aa 2> /dev/null
+agi*
+EOF
+EXPECT=<<EOF
+agn "sym.imp.free" base64:
```
Comment (ret2libc) [inline test/db/cmd/cmd_ag:8]: master currently doesn't print the empty string. Which, according to the help of `agn` seems wrong because a body is expected. However currently `agn <title> base64:` fails.

File: libr/core/agraph.c (hunk 5 above)
Comment (karliss) [reply libr/core/agraph.c:4942]: TLDR: `need_reload_nodes` causes agraph to attempt loading either callgraph or basic block graph, which doesn't make sense in this case. 

Because agraph is mixing multiple responsibilities within single module. On hand it deals with task of making exact drawing on other it also handles reloading of data from r2 for some kinds of graphs like function CFG. Many of agraph properties are specific to one kind of graph.

File: test/db/cmd/cmd_ag [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+.--------------------.
+|  main              |
+`--------------------'
+digraph code {
+rankdir=LR;
+outputorder=edgesfirst
+graph [bgcolor=azure];
+edge [arrowhead=normal, color="#3030c0" style=bold weight=2];
+node [fillcolor=white, style=filled shape=box fontname="Courier" fontsize="8"];
+0 [URL="sym.imp.free", color="lightgray", label="sym.imp.free"]
```
Comment (ret2libc) [inline test/db/cmd/cmd_ag:79]: master here uses "sym.imp.free" instead of 0. Is it ok?

**Review by ret2libc — COMMENTED** [review]
So, if I get it right, the final purpose is to also move all those other things that currently work on `core->graph` directly in `r_core_agraph_print` to `r_core_graph_print` and make somehow core->graph be a graph of RGraphNodeInfo, right?

Would it be already possible to rewrite `r_core_agraph_print` in terms of `r_core_graph_print` or do you prefer to do it in a separate PR?

Anyway, the general idea looks ok to me. However, as said previously, I think `RGraphNodeInfo` doesn't really belong to `r_util/r_graph.h`. I'd prefer to have a separate header for it `r_graph_drawable` or something, just for clear separation.

File: libr/util/graph.c (hunk 2 above)
Comment (karliss) [reply libr/util/graph.c:170]: Yes I was having some doubts on which of the two strategies to choose. Here are the reasons I chose this strategy:
* "This function argument takes ownership, ALWAYS!" seemed like a simpler rule than "It takes ownership but sometimes it doesn't"
* Less error handling code and cleaner code in function using the API to build a graph.
* Handling of the case when construction consists of two steps and first step succeeds transfering the ownership, but second step fails. Here is an example to make the problem clearer:
```c
foo_bar *make_foobar(char *a)
{
    foo * foo_obj = make_foo(a);
    if (!foo_obj) {
       return NULL;
   }
   foo_x(foo_obj);
   foo_bar * foo_bar_obj = malloc(...);
   if (!foo_bar) {
       // What to do now? Ownership of a was already transfered to foo_obj. Can't call the normal destructor of foo, because that would also free a. What if foo_x or even make_foo can modify or replace `a`?
   }
   return foo_bar_obj;
}
```

It seems that neither of approaches is widely used and in most case the approach you suggested means that no one in case of erorr memory is leaked because caller also doesn't release it.  If you consider that my approach is likely to cause confusion I would prefer the third option of not doing the ownership transfer at all and doing a copy when necessary. That way at least there is less conditional code on most levels.

File: libr/core/agraph.c (hunk 5 above)
Comment (ret2libc) [reply libr/core/agraph.c:4942]: ok! ;)

**Comment by karliss** [comment]
> So, if I get it right, the final purpose is to also move all those ...

Most cases except the cases maybe some cases that do custom stuff in a way that provides additional value. Replacing core->graph with RGraph<RGraphNodeInfo>+dictionary<string, node> would be one of the last steps. Yes I would preffer doing r_core_agraph_print rewrite separately. I hope that doing things in multiple smaller PRs will make both review process and further refactoring easier once there is less code directly depending core->graph. I expect there to be a bit of extra work to untangle the parts important for `agg` from `afg` specific code.

An approximate plan is:
*) Initial PR (this one), introduce the concept, demonstrate it in one or two cases. During review debate various design choices. Don't try to convert everything so that in case review requires rewriting some parts only the 1-3 examples need to be redone.
*) Convert most of the cases that don't need special handling. Convert the cases that currently have special handling but don't need it.
*) agg refactoring, reduce code duplication between mechanism I introduced and agg. See what can be cleaned up within agraph.

I will move the relevant code from r_graph to r_graph_drawable within this PR.

File: test/db/cmd/cmd_ag (hunk 11 above)
Comment (karliss) [reply test/db/cmd/cmd_ag:8]: I hav

Comment (karliss) [reply test/db/cmd/cmd_ag:8]: No problems removing it. The behavior in current master is a bit inconsistent. If you do `agi*; .agi*; agg*` first one prints without `base64:` but last one prints it.

File: libr/util/graph.c (hunk 2 above)
Comment (ret2libc) [reply libr/util/graph.c:170]: Probably both approaches have their own pros/cons, but so far I think we are used to the approach "take ownership only if there's no error". At least, this is what I think is used in r_list and other data structures. @thestr4ng3r @trufae opinions?

File: libr/core/cmd_anal.c (hunk 9 above)
Comment (karliss) [reply libr/core/cmd_anal.c:8887]: What's exactly you suggestion? Keep spaces for this block or reindent it whole?

Comment (ret2libc) [reply libr/core/cmd_anal.c:8887]: Oh damn... it's all messed up... I thought this was "the wrong one", instead this is the right one and all others are wrong :D I would say to just make it wrong as well :PP

File: test/db/cmd/cmd_ag (hunk 12 above)
Comment (karliss) [reply test/db/cmd/cmd_ag:79]: Why not? Its valid dot file. Less copies of the the same string, less place in code that needs to properly print those strings. Allow this function to be used with graphs that don't have unique titles.

File: libr/core/cmd_anal.c (hunk 10 above)
Comment (karliss) [reply libr/core/cmd_anal.c:8107]: Fixed

File: libr/core/canal.c [hunk 13]
```diff
@@ -2280,43 +2280,67 @@ R_API void r_core_anal_coderefs(RCore *core, ut64 addr) {
...
+	RGraphNode *curr_node;
+	char *me;
+	if (!graph) {
+		return;
+	}
+	me = (f && f->offset == addr)
+		? r_str_new (f->name)
+		: r_str_newf ("0x%" PFMT64x, addr);
+
+	curr_node = r_graph_add_node_info (graph, me, NULL, addr);
```
Comment (trufae) [inline libr/core/canal.c:2294]: define and assign in the same line

File: libr/core/canal.c [hunk 14]
```diff
@@ -2280,43 +2280,67 @@ R_API void r_core_anal_coderefs(RCore *core, ut64 addr) {
 	}
 }
 
-R_API void r_core_anal_importxrefs(RCore *core) {
+static void add_single_addr_xrefs(RCore *core, ut64 addr, RGraph *graph) {
+	RFlagItem *f = r_flag_get_at (core->flags, addr, false);
+	RGraphNode *curr_node;
+	char *me;
+	if (!graph) {
```
Comment (trufae) [inline libr/core/canal.c:2287]: use r_return_if_fail

File: libr/core/canal.c [hunk 15]
```diff
@@ -2280,43 +2280,67 @@ R_API void r_core_anal_coderefs(RCore *core, ut64 addr) {
...
+	bool lit = info? info->has_lit: false;
 	bool va = core->io->va || core->bin->is_debugger;
 
 	RListIter *iter;
 	RBinImport *imp;
 	if (!obj) {
-		return;
+		return NULL;
+	}
+	graph = r_graph_new ();
```
Comment (trufae) [inline libr/core/canal.c:2322]: define and assign in the same line

File: libr/util/graph.c [hunk 16]
```diff
@@ -316,6 +328,23 @@ R_API RGraphNodeInfo *r_graph_create_node_info(char *title, char *body, ut64 off
 		data->title = title;
 		data->body = body;
```
Comment (trufae) [inline libr/util/graph.c:329]: i think this is a setter and it should be responsible for owning the passed string, so better use const char* for body and title. just strdup in here and remove the frees below

File: libr/util/graph.c [hunk 17]
```diff
@@ -305,6 +314,9 @@ R_API void r_graph_dfs(RGraph *g, RGraphVisitor *vis) {
 
 R_API void r_graph_free_node_info(void *ptr) {
 	RGraphNodeInfo *info = ptr;
+	if (!info) {
```
Comment (trufae) [inline libr/util/graph.c:317]: if (!ptr) one line above?

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/util/graph.c (hunk 2 above)
Comment (ret2libc) [reply libr/util/graph.c:170]: BTW, see also https://github.com/radareorg/radare2/pull/17621/files#diff-635a23ca710c8cf67d67f28d75cdcf86R24 . I implemented `ht_`, so I'm not sure how much that qualifies as a valid example, but even there, if `n` is not correctly inserted in the ht, it is still freed by the caller.

File: libr/core/cmd_anal.c [hunk 18]
```diff
@@ -8521,7 +8708,7 @@ static void cmd_anal_graph(RCore *core, const char *input) {
 		case 'w':
 		case ' ': {
 			core->graph->is_callgraph = true;
-			r_core_cmdf (core, "ag-; .agc* @ %"PFMT64u"; agg%s;", core->offset, input + 1);
+			r_core_cmdf (core, "ag-; .agc* @ %" PFMT64u "; agg%s;", core->offset, input + 1);
```
Comment (trufae) [inline libr/core/cmd_anal.c:8711]: Code injection

File: libr/core/canal.c [hunk 19]
```diff
@@ -2280,43 +2281,65 @@ R_API void r_core_anal_coderefs(RCore *core, ut64 addr) {
 	}
 }
 
-R_API void r_core_anal_importxrefs(RCore *core) {
+static void add_single_addr_xrefs(RCore *core, ut64 addr, RGraph *graph) {
+	RFlagItem *f = r_flag_get_at (core->flags, addr, false);
+	char *me;
+	r_return_if_fail (graph);
+	me = (f && f->offset == addr)
```
Comment (trufae) [inline libr/core/canal.c:2288]: Define and assign in the same line

File: libr/core/cmd_anal.c [hunk 20]
```diff
@@ -8286,25 +8324,186 @@ R_API void r_core_agraph_print (RCore *core, int use_utf, const char *input) {
...
+		}
+		break;
+	}
+	case '*': // "ag_*" -
+		print_graph_agg (graph);
+		break;
+	case 'J':
+	case 'j': {
+		PJ *pj = pj_new ();
+		if (!pj) {
```
Comment (trufae) [inline libr/core/cmd_anal.c:8469]: I would invert the check and remove the return

File: libr/util/graph_drawable.c [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+R_API RGraphNode *r_graph_add_node_info(RGraph *graph, const char *title, const char *body, ut64 offset) {
+	r_return_val_if_fail (graph, NULL);
+	RGraphNodeInfo *data = r_graph_create_node_info (title, body, offset);
+	if (!data) {
+		return NULL;
+	}
+	return r_graph_add_nodef (graph, data, r_graph_free_node_info);
+}
+
+char *r_graph_drawable_to_dot (RGraph /*RGraphNodeInfo*/ *graph, const char *node_properties, const char *edge_properties) {
```
Comment (trufae) [inline libr/util/graph_drawable.c:33]: If its public use R_API

Also im not 100% happy eith the drawable name but im fine for all the changes proposed here

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_anal.c (hunk 18 above)
Comment (karliss) [reply libr/core/cmd_anal.c:8711]: *) was there before this issue
*) might get removed in the follow up issues which convert more graphs to the mechanism introduce in this issue

File: libr/util/graph_drawable.c (hunk 21 above)
Comment (karliss) [reply libr/util/graph_drawable.c:33]: Do you have suggestions for better name?

**Review by XVilka — APPROVED** [review]
Looks very good to me!

File: libr/core/cmd_anal.c [hunk 22]
```diff
@@ -8258,14 +8295,15 @@ R_API void r_core_agraph_print (RCore *core, int use_utf, const char *input) {
 	case 'd': { // "aggd" - dot format
 		const char *font = r_config_get (core->config, "graph.font");
 		r_cons_printf ("digraph code {\nrankdir=LR;\noutputorder=edgesfirst\ngraph [bgcolor=azure];\n"
-			"edge [arrowhead=normal, color=\"#3030c0\" style=bold weight=2];\n"
-			"node [fillcolor=white, style=filled shape=box "
-			"fontname=\"%s\" fontsize=\"8\"];\n", font);
+			       "edge [arrowhead=normal, color=\"#3030c0\" style=bold weight=2];\n"
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8298]: no need to change these lines, they were right as they were before AFAIK.

File: libr/util/graph.c (hunk 2 above)
Comment (ret2libc) [reply libr/util/graph.c:170]: You said this reply was not required any more. So, is this going away? Are changes still missing from this PR?

File: libr/include/r_util/r_graph_drawable.h [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+	 * @brief Optional offset for the object corresponding to node.
+	 * 
+	 * Interactive output modes can use it to provide actions like seeking to
+	 * this position or modify the object.
+	 */
+	ut64 offset;
+} RGraphNodeInfo;
+
+R_API void r_graph_free_node_info(void *ptr);
+R_API RGraphNodeInfo *r_graph_create_node_info(const char *title, const char *body, ut64 offset);
```
Comment (ret2libc) [inline libr/include/r_util/r_graph_drawable.h:30]: I guess we can start by hiding this from the API as it's not used by any module outside the r_graph_add_node_info function.

File: libr/util/graph.c (hunk 2 above)
Comment (karliss) [reply libr/util/graph.c]: Missed that one due to it being separate file. Removed.

File: libr/util/graph_drawable.c (hunk 21 above)
Comment (karliss) [reply libr/util/graph_drawable.c]: Adding missing R_API to function definitions.

File: libr/core/cmd_anal.c (hunk 22 above)
Comment (karliss) [reply libr/core/cmd_anal.c]: Clang format decided to change that, reverted.

**Comment by XVilka** [comment]
@ret2libc please take a look again and lets merge this.

File: libr/core/cmd_anal.c [hunk 24]
```diff
@@ -8286,25 +8324,186 @@ R_API void r_core_agraph_print (RCore *core, int use_utf, const char *input) {
...
+
+static void print_graph_agg(RGraph /*RGraphNodeInfo*/ *graph) {
+	RGraphNodeInfo *print_node;
+	RGraphNode *node, *target;
+	RListIter *it, *edge_it;
+	r_list_foreach (graph->nodes, it, node) {
+		char *encbody;
+		int len;
+		print_node = node->data;
+		if (print_node->body && print_node->body[0]) {
```
Comment (ret2libc) [inline libr/core/cmd_anal.c:8357]:
```suggestion
		if (R_STR_ISNOTEMPTY (print_node->body)) {
```

File: libr/util/graph_drawable.c [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+		return NULL;
+	}
+	RGraphNode *node = r_graph_add_nodef (graph, data, r_graph_free_node_info);
+	if (!node) {
+		r_graph_free_node_info (data);
+	}
+	return node;
+}
+
+R_API char *r_graph_drawable_to_dot (RGraph /*RGraphNodeInfo*/ *graph, const char *node_properties, const char *edge_properties) {
```
Comment (ret2libc) [inline libr/util/graph_drawable.c:37]:
```suggestion
R_API char *r_graph_drawable_to_dot(RGraph /*RGraphNodeInfo*/ *graph, const char *node_properties, const char *edge_properties) {
```

File: libr/util/graph_drawable.c [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+		}
+		r_list_foreach (node->out_nodes, itt, target) {
+			r_strbuf_appendf (&buf, "%d -> %d\n", node->idx, target->idx);
+		}
+	}
+	r_strbuf_append (&buf, "}\n");
+	return r_strbuf_drain_nofree (&buf);
+}
+
+R_API void r_graph_drawable_to_json (RGraph /*RGraphNodeInfo*/ *graph, PJ *pj, bool use_offset) {
```
Comment (ret2libc) [inline libr/util/graph_drawable.c:69]:
```suggestion
R_API void r_graph_drawable_to_json(RGraph /*RGraphNodeInfo*/ *graph, PJ *pj, bool use_offset) {
```

**Comment by ret2libc** [comment]
Please address opened comments by pancake and me, but overall I'm ok with the PR. :+1:

File: libr/core/canal.c (hunk 19 above)
Comment (karliss) [reply libr/core/canal.c]: I am fine with whatever style you choose, but shouldn't code style guide be updated in that case. https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md requests not to use C99 variable declarations.

**Review by XVilka — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]

**Comment by XVilka** [comment]
One unit test is still failing: https://github.com/radareorg/radare2/pull/17631/checks?check_run_id=1205972042#step:13:825

File: libr/anal/class.c [hunk 27]
```diff
@@ -2,6 +2,7 @@
 
 #include <r_anal.h>
 #include <r_vector.h>
+#include <r_util/r_graph_drawable.h>
```
Comment (trufae) [inline libr/anal/class.c:5]: this file is not in libr/meson.build. this is why ci is red.

**Review by trufae — CHANGES_REQUESTED** [review]
